### PR TITLE
Make sure top-level callback has correct context

### PR DESCRIPTION
**Pull request by spicyj**
Closes #5125 -- incorporates essentially the same fix with a much simpler test case. Thanks @yiminghe for reporting!

### CODE REVIEW

**Comment by yiminghe** [comment]
You can not get public instance
```
publicInst = prevComponent._renderedComponent.getPublicInstance();
```

outside updatedCallback, because publicInstance may change, so You have to get public instance inside updatedCallback:

```js
var updatedCallback = callback && function() {
  callback.call(prevComponent._renderedComponent.getPublicInstance());
};
```

You can test on my tc.

File: src/renderers/dom/client/ReactMount.js [hunk 1]
```diff
@@ -609,12 +609,17 @@ var ReactMount = {
       var prevWrappedElement = prevComponent._currentElement;
       var prevElement = prevWrappedElement.props;
       if (shouldUpdateReactComponent(prevElement, nextElement)) {
-        return ReactMount._updateRootComponent(
+        var publicInst = prevComponent._renderedComponent.getPublicInstance();
```
Comment (yiminghe) [inline src/renderers/dom/client/ReactMount.js]: publicInstance must be get inside updatedCallback.

see tc from https://github.com/facebook/react/pull/5125/files

**Comment by spicyj** [comment]
Your test case already passes with my code here. You can see that in my test case I tested both the case that the type changes (and you get a new instance) and when it doesn't. The public instance for an internal instance never changes, at least with our current design.

File: src/renderers/dom/client/ReactMount.js [hunk 2]
```diff
@@ -610,12 +610,17 @@ var ReactMount = {
...
+        var updatedCallback = callback && function() {
```
Comment (zpao) [inline src/renderers/dom/client/ReactMount.js]: Can we just `.bind(publicInst)`?

**Comment by yiminghe** [comment]
You're right, I just noticed top level wrapper is different from normal ReactCompositeComponent, top level wrapper will compare `wrapper._currentElement.props` directly...

**Pull request merged**